inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp

```diff
@@ -427,25 +429,32 @@ void kernels_cache::build_all() {
     if (!_pending_compilation)
         return;
 
-    std::lock_guard<std::mutex> lock(_context.get_cache_mutex());
-
     auto sorted_program_code = get_program_source(_kernels_code);
+    {
+        std::lock_guard<std::mutex> lock(_context.get_cache_mutex());
+        _one_time_kernels.clear();
+    }
 
-    _one_time_kernels.clear();
+    std::vector<std::future<void>> builds;
     for (auto& program : sorted_program_code) {
-        auto kernels = build_program(program.second);
-
-        for (auto& k : kernels) {
-            const auto& entry_point = k.first;
-            const auto& k_id = program.second.entry_point_to_id[entry_point];
-            if (program.second.one_time) {
-                _one_time_kernels[k_id] = k.second;
-            } else {
-                _kernels[k_id] = k.second;
+        builds.push_back(std::async(std::launch::async,[&]() {
+            auto kernels = build_program(program.second);
```

Comment:
Does it work correctly with cl_cache? Сompute runtime [FAQ](https://github.com/intel/compute-runtime/blob/master/opencl/doc/FAQ.md#what-are-the-known-limitations-of-cl_cache) says:
```
What are the known limitations of cl_cache?
Not thread safe. (Workaround: Make sure your clBuildProgram calls are executed in thread safe fashion.)
```

Reply:
It is unclear from the FAQ that you are referring to.

clBuildProgram itself is thread-safe (https://community.khronos.org/t/are-clcreateprogramwithsource-clbuildprogram-thread-safe/2771) since OCL 1.1, also the build_program function is marked as const and has no mutable objects (so it doesn't change the internal state)...so I guess the question is whether cl_cache **update** (with the clBuildProgram/clCreateKernel) results is thread-safe.

That is why the mutex now moved to protect the corresponding std::map for the kernels/one_time_kernels.
If in doubts..let's redirect to the people that originally developed the kernels_cache code


The same FAQ states that the process-wise is not safe the cache upadte

Reply:
Most likely cl_cache requires thread-safe `clBuildProgram` call due to work with the file system to make sure that there are no data races for read/write of the binary files. Since cl_cache is not a part of OCL standard, it might introduce additional limitations that are not described in the spec. So I think we need an explicit confirmation from compute runtime team that this code is safe.

Reply:
Again, clBuildProgram is thread-safe per spec, so you pointed just right - what's really needs to be protected are bin files.
As our code builds different programs/kernels in parallel (and saves to the cl_cache), this shouldn't be an issue.
Also, saving/loading is guarded at the app level with the additional mutex anyway:
https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp#L71
https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp#L100
This is not touched.

My change actually makes protection of the clDNN's **internal** structures more fine-grained.
Let's add @mletavin as we was also contributing to the file before

Reply:
> As our code builds different programs/kernels in parallel (and saves to the cl_cache)

What if we enable cl_cache and load MULTI plugin for 2 dg1 cards? Kernels will be the same in both programs and they can be reused, and my point is that it **might** cause some issues with parallel build enabled.

mutex for binary files load/store is fine, but it works only for **plugin's cache** (enabled via CACHE_DIR IE config option) which is not the same as **driver's cache** (enabled via cl_cache_dir env variable)
So I don't see any issues for plugin's cache for now, but I'm not sure that driver's cache will work fine. 

Technically, there is similar mutex in the driver for bin files read/write: https://github.com/intel/compute-runtime/blob/master/shared/source/compiler_interface/compiler_cache.cpp#L69 
So given that clBuildProgram call is thread-safe and operations with bin files inside ocl runtime are protected with mutex, then there's a good chance that it will also work fine (at least with new driver version). 

@AdamCetnerowski could you clarify if we can safely call clBuildProgram in parallel when cl_cache is enabled?

Reply:
I suggest to introduce new plugin flag and put this parallel compilation under it.
This will move responsibility of enabling it to the user, which perfectly coincides with **cl_cache** ideology.
So, user will have means to either disable parallel compilation, if he wants to generate clean cache, or enable both to get maximum speed.
In my opinion, such an approach will remove any contradiction of parallel compilation and **cl_cache**.

Reply:
per discussion over e-mail, we will discourage users from the cl_cache, until it becomes safe to use, and invest into OV (and GPU plugin cache in particular)